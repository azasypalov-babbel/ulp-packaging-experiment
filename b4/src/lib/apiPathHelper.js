export const buildPath = ({ version, locale, uuid, learnLanguageAlpha3 }, rest = '') => {
  if (!version) {
    throw 'Missing required param "version"';
  }
  if (!locale) {
    throw 'Missing required param "locale"';
  }
  if (!uuid) {
    throw 'Missing required param "uuid"';
  }
  if (!learnLanguageAlpha3) {
    throw 'Missing required param "learnLanguageAlpha3"';
  }

  return `v${version}/${locale}/accounts/${uuid}/learn_languages/${learnLanguageAlpha3}/${rest}`;
};

export const buildPathV5 = ({ version, locale, uuid, contentReleaseId, learnLanguageAlpha3 }, rest = '') => {
  if (!version) {
    throw 'Missing required param "version"';
  }
  if (!locale) {
    throw 'Missing required param "locale"';
  }
  if (!uuid) {
    throw 'Missing required param "uuid"';
  }
  if (!learnLanguageAlpha3) {
    throw 'Missing required param "learnLanguageAlpha3"';
  }
  if (!contentReleaseId) {
    throw 'Missing required param "contentReleaseId"';
  }

  return `v${version}/${locale}/accounts/${uuid}/content_releases/${contentReleaseId}` +
    `/learn_languages/${learnLanguageAlpha3}/${rest}`;
};
