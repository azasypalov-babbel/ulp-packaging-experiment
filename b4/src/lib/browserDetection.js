// please note,
// that IE11 now returns undefined again for window.chrome
// and new Opera 30 outputs true for window.chrome
// but needs to check if window.opr is not undefined
// and new IE Edge outputs to true now for window.chrome
// and if not iOS Chrome check
// so use the below updated condition

// based on https://stackoverflow.com/a/13348618
const isChromium = window.chrome;
const agent = window.navigator.userAgent.toLowerCase();
const vendorName = window.navigator.vendor;
const isOpera = typeof window.opr !== 'undefined';
const isIEedge = agent.indexOf('edge') > -1 || agent.indexOf('edg') > -1;
let isChrome = false;

if (
  isChromium !== null &&
  typeof isChromium !== 'undefined' &&
  vendorName === 'Google Inc.' &&
  isOpera === false &&
  isIEedge === false
) {
  isChrome = true;
}

export { isChrome };
