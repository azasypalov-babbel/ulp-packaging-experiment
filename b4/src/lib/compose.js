export const compose = (...functionsToCompose) => (argumentsOfFunctions) => {
  return functionsToCompose.reduceRight((result, functionToCompose) => {
    return functionToCompose(result);
  }, argumentsOfFunctions);
};
