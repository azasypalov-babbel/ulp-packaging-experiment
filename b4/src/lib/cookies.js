function identity(x) {
  return x;
}

function convertFeatureStrToObject(elem) {
  const elements = elem.trim().split('=');
  return {
    [elements[0]]: elements[1]
  };
}

function arrayToObjectReducer(acc, element) {
  return {
    ...acc,
    ...element
  };
}

function getCookies() {
  return document?.cookie.split(';')
    .filter(identity)
    .map(convertFeatureStrToObject)
    .reduce(arrayToObjectReducer, {});
}

function getCookie(cookie) {
  const cookies = getCookies();

  return cookies && cookies[cookie];
}

function setCookie(key, value, expiryDate) {
  document.cookie = `${key}=${value};expires=${expiryDate.toUTCString()};path=/`;
}

export default {
  setCookie,
  getCookies,
  getCookie
};
