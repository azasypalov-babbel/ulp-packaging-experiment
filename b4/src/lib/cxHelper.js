import baseCX from 'classnames';

const PREFIX = 'loy';

const cx = (...args) => {
  const classNames = baseCX(args);

  return classNames.split(' ')
    .map((className) => `${PREFIX}-${className}`)
    .join(' ');
};

export default cx;
