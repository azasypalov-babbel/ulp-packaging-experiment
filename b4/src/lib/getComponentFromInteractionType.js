import Fillin from '../components/Interactions/Fillin/FillinInteraction';
import Puzzlehelper from '../components/Interactions/Puzzlehelper/PuzzlehelperInteraction';
import Wordorder from '../components/Interactions/Wordorder/WordorderInteraction';
import ChoiceButtons from '../components/Interactions/Choicebuttons/ChoicebuttonsInteraction';
import Speak from '../components/Interactions/Speak/SpeakInteraction';

const interactionMap = {
  speak: Speak,
  fillin: Fillin,
  puzzlehelper: Puzzlehelper,
  wordorder: Wordorder,
  choose: ChoiceButtons
};

const getComponentFromInteractionType = (interactionType) => interactionMap[interactionType];

export default getComponentFromInteractionType;

