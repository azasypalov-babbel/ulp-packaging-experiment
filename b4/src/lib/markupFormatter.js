import { compose } from '../lib/compose';
import {
  formatParentheses,
  getFirstCorrectSolution,
  removeGapFormatting,
  removeFirstParanthesesAndGapsFormatting,
  markupStringToPlainText
} from '@lessonnine/babbel-markup-helper.js';

const commonFormat = compose(
  removeGapFormatting,
  formatParentheses,
  markupStringToPlainText
);

export const formatDL = (s) => {
  if (!s) return '';

  return compose(
    removeFirstParanthesesAndGapsFormatting,
    commonFormat
  )(s);
};

export const formatLL = (s) => {
  if (!s) return '';

  return compose(
    commonFormat,
    getFirstCorrectSolution
  )(s);
};

export const stripParantheses = (text) => text.replace(/(\([\S]+\))/g, '').trim();

