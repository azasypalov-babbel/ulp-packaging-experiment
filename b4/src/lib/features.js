import cookies from './cookies';
const { setCookie, getCookies } = cookies;

let featureStore = {};

const defaultAllowedFeatures = [
  'is_demo',
  'is_redux_logger',
  'is_redux_devtools',
  'is_unlocked',
  'is_cat_preview',
  'is_analyzer_lib_recognizer',
  'is_b3_vocabulary_trainer',
  'is_web_speech_mock',
  'is_native_speech',
  'is_verbose_logging',
  'is_transliteration',
  'is_fake_caret',
  'is_selection_helper',
  'is_puzzle_helper',
  'is_debug',
  'is_refer_a_friend',
  'is_zendesk'
];

const defaultNonPersistentFeatures = [
  'is_demo',
  'is_analyzer_lib_recognizer',
  'is_b3_vocabulary_trainer',
  'is_fake_caret',
  'is_selection_helper',
  'is_puzzle_helper',
  'is_zendesk'
];

function filterFeatures(features, allowedFeatures) {
  const filteredFeatures = {};
  Object.keys(features).forEach(function(key) {
    if (allowedFeatures.includes(key) && (features[key] === 'on' || features[key] === 'off')) {
      filteredFeatures[key] = features[key];
    }
  });
  return filteredFeatures;
}

function filterPersistentFeatures(features, nonPersistentFeatures) {
  const filteredFeatures = {};
  Object.keys(features).forEach(function(key) {
    if (!nonPersistentFeatures.includes(key)) {
      filteredFeatures[key] = features[key];
    }
  });
  return filteredFeatures;
}

export function configure(
  features,
  allowedFeatures = defaultAllowedFeatures,
  nonPersistentFeatures = defaultNonPersistentFeatures
) {
  featureStore = filterFeatures(features, allowedFeatures);

  const persistantFeatures = filterPersistentFeatures(
    filterFeatures(getCookies(), allowedFeatures),
    nonPersistentFeatures
  );

  Object.entries(persistantFeatures)
    .forEach(function([key, value]) {
      if (!(key in featureStore)) {
        featureStore[key] = value;
      }
    });

  const expiryDate = new Date();
  const oneYearMs = 365 * 24 * 60 * 60 * 1000;
  expiryDate.setTime(expiryDate.getTime() + oneYearMs);

  Object.entries(filterPersistentFeatures(featureStore, nonPersistentFeatures))
    .forEach(function([key, value]) {
      setCookie(key, value, expiryDate);
    });
}

export function get(featureName) {
  if (featureStore[featureName] === 'on') {
    return true;
  } else if (featureStore[featureName] === 'off') {
    return false;
  } else {
    return null;
  }
}

export function getActiveFeatures() {
  return Object.keys(featureStore)
    .filter((key) => featureStore[key] === 'on');
}

export function isWebview() {
  return process.env.WEBVIEW === '1';
}

export function isAndroid() {
  return (
    isWebview()
    && typeof window.androidListener?.postMessage === 'function'
  );
}

export function isiOS() {
  return (
    isWebview() &&
    typeof window.webkit?.messageHandlers?.iosListener?.postMessage === 'function'
  );
}
