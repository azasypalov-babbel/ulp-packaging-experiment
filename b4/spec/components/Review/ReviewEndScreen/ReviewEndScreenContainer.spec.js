import {
  mapStateToProps,
  mergeProps,
  getGrade
} from '../../../../src/components/Review/ReviewEndScreen/ReviewEndScreenContainer';

import { formatLL } from '../../../../src/lib/markupFormatter';
jest.mock('../../../../src/lib/markupFormatter', () => ({
  formatLL: jest.fn((arg) => arg)
}));

const mockServices = {
  soundService: jest.fn(),
  mediaUrlService: jest.fn()
};

const mockItems = [
  { learnLanguageText: 'a', sound: { id: '100afffb3' } },
  { learnLanguageText: 'b', sound: { id: '101afffb3' } },
  { learnLanguageText: 'c', sound: { id: '102afffb3' } }
];

jest.mock('../../../../src/dux/sequence/selectors', () => ({
  purgeableItems: () => [{}, {}, {}],
  correctItems: () => [mockItems[0], mockItems[1]],
  incorrectItems: () => [mockItems[2]]
}));

describe('ReviewEndScreenContainer', () => {
  const props = {
    ...mockServices,
    onCorrectErrorsButtonClick: jest.fn(),
    onReviewMoreButtonClick: jest.fn(),
    onGetMoreInfoLinkClick: jest.fn(),
    onCloseButtonClick: jest.fn(),
    showCloseButton: false
  };
  const state = {
    account: {
      data: {
        displayname: 'Julia'
      }
    },
    session: {
      locale: 'locale-test'
    },
    review: {
      hasNextReviewItems: false,
      reviewItems: {
        vocabulary: {
          trainers: []
        }
      }
    },
    sequence: {}
  };

  describe('#mapStateToProps', () => {
    it('should return the correct props', () => {
      expect(mapStateToProps(state)).toMatchSnapshot();
    });
  });

  describe('#mergeProps', () => {
    it('should return the correct props', () => {
      const stateProps = {
        purgeableItemsCount: 3,
        correctItems: [mockItems[0], mockItems[1]],
        incorrectItems: [mockItems[2]],
        displayName: 'Julia'
      };
      expect(mergeProps(stateProps, null, props)).toMatchSnapshot();
    });
  });

  describe('#getGrade', () => {
    it('should return correct grade', () => {
      expect(getGrade(0, 10)).toEqual('low');
      expect(getGrade(5, 10)).toEqual('medium');
      expect(getGrade(10, 10)).toEqual('high');
    });
  });

  describe('learn language text', () => {
    beforeEach(() => {
      formatLL.mockClear();
      mapStateToProps(state);
    });

    test('calls formatLL to remove CAT formatting', () => {
      expect(formatLL).toHaveBeenCalledTimes(mockItems.length);
    });
  });
});
