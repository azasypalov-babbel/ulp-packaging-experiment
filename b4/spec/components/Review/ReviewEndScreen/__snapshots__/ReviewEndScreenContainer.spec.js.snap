// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewEndScreenContainer #mapStateToProps should return the correct props 1`] = `
Object {
  "correctItems": Array [
    Object {
      "learnLanguageText": "a",
      "sound": Object {
        "id": "100afffb3",
      },
    },
    Object {
      "learnLanguageText": "b",
      "sound": Object {
        "id": "101afffb3",
      },
    },
  ],
  "displayName": "Julia",
  "hasNextReviewItems": false,
  "incorrectItems": Array [
    Object {
      "learnLanguageText": "c",
      "sound": Object {
        "id": "102afffb3",
      },
    },
  ],
  "locale": "locale-test",
  "purgeableItemsCount": 3,
}
`;

exports[`ReviewEndScreenContainer #mergeProps should return the correct props 1`] = `
Object {
  "correctItems": Array [
    Object {
      "learnLanguageText": "a",
      "sound": Object {
        "id": "100afffb3",
      },
    },
    Object {
      "learnLanguageText": "b",
      "sound": Object {
        "id": "101afffb3",
      },
    },
  ],
  "displayName": "Julia",
  "grade": "high",
  "incorrectItems": Array [
    Object {
      "learnLanguageText": "c",
      "sound": Object {
        "id": "102afffb3",
      },
    },
  ],
  "mediaUrlService": [MockFunction],
  "onCloseButtonClick": [MockFunction],
  "onCorrectErrorsButtonClick": [MockFunction],
  "onGetMoreInfoLinkClick": [MockFunction],
  "onReviewMoreButtonClick": [MockFunction],
  "playEndScreenSound": [Function],
  "playSound": [Function],
  "purgeableItemsCount": 3,
  "showCloseButton": false,
  "soundService": [MockFunction],
  "totalItemsCount": 3,
}
`;
